feat: Job Masters and Job Training

pacampbell opened the pull request:
- Added support for Job Masters.
- All custom skills above level 4 are rolled back to level 4
- All augments above level 3 are rolled back to level 3
- Skills which are rolled back have their job points refunded to the
  player.
- All skills and augments are unequipped from all players.

Miscellaneous Changes:
- Implemented the personal quest "The High Scepter's Heir" which is used
  to unlock High Scepter and High Scepter job training.
- Adjusted automatic EXP and BO calculations.

# Checklist:
- [x] The project compiles
- [x] The PR targets `develop` branch

pacampbell committed:
feat: Job Masters and Job Training

- Added support for Job Masters.
- All custom skills above level 4 are rolled back to level 4
- All augments above level 3 are rolled back to level 3
- Skills which are rolled back have their job points refunded to the
  player.
- All skills and augments are unequipped from all players.

Miscellaneous Changes:
- Implemented the personal quest "The High Scepter's Heir" which is used
  to unlock High Scepter and High Scepter job training.
- Adjusted automatic EXP and BO calculations.

pacampbell commented on Arrowgene.Ddon.Shared/Model/JobId.cs:
```diff
@@ -59,5 +59,24 @@ public static bool IsBlueJob(this JobId jobId)
+            [JobId.Warrior] = ContentsRelease.WarriorJobTraining,
+            [JobId.ElementArcher] = ContentsRelease.ElementArcherJobTraining,
+            [JobId.Alchemist] = ContentsRelease.AlchemistJobTraining,
+            [JobId.SpiritLancer] = ContentsRelease.SpiritLancerJobTraining,
```
Missing HS

RyanYappert commented on Arrowgene.Ddon.GameServer/Characters/JobMasterManager.cs:
```diff
@@ -0,0 +1,217 @@
+
+            bool isOrbEnemy = enemy.BloodOrbs > 0;
+
+            foreach (var memberClient in client.Party.Clients)
```
This is an odd place to put this iteration, since this is called from `InstanceEnemyKillHandler` which already has an iteration over the party members.

RyanYappert commented on Arrowgene.Ddon.GameServer/Handler/InstanceEnemyKillHandler.cs:
```diff
@@ -132,6 +132,8 @@ public override S2CInstanceEnemyKillRes Handle(GameClient client, C2SInstanceEne
                     }
                 }
 
+                queuedPackets.AddRange(Server.JobMasterManager.HandleEnemyKill(client, enemyKilled, connectionIn));
```
Again, an odd place to put this, because you're now 1) circumventing the IsNoBattleReward check (which otherwise works on hunt quests and achievements and other tracking) and 2) circumvents the proximity check, so you can be in WDT while somebody else does your job training for you.

RyanYappert replied:
Where I'd put it would be after 
```
queuedPackets.AddRange(Server.AchievementManager.HandleKillEnemy(memberClient, enemyKilled, connectionIn: connectionIn));
``` 
around the hunt quest/BO/HO/PP/EXP/Achievement handling.

pacampbell committed:
Address code review feedback

- Adjusted how enemy kill detection works

pacampbell committed:
fixes: Small fixes based on testing

Fixed an issue where training orders could be received and progressed
before unlocking job training for that particular job.

pacampbell commented:
/build
